{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "21f4c62c",
   "metadata": {},
   "source": [
    "# PySam/Bamnostic: Navigating through a SAM/BAM file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8cb4d573",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n# Install a pip package in the current Jupyter kernel\\nimport sys\\n!{sys.executable} -m pip install bamnostic\\n'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "# Install a pip package in the current Jupyter kernel\n",
    "import sys\n",
    "!{sys.executable} -m pip install bamnostic\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "afb670d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bamnostic as bs \n",
    "import pysam\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "from collections import defaultdict "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "25d7aa05",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " UserWarning:/home/prabakaranjd/.local/lib/python3.8/site-packages/bamnostic/bam.py:310: No supplied index file was not found. Random access disabled\n"
     ]
    }
   ],
   "source": [
    "# Sorted BAM files\n",
    "\n",
    "p1mat = '1mat_sort.bam'\n",
    "p2mat = '2mat_sort.bam'\n",
    "p1pat = '1pat_sort.bam'\n",
    "p2pat = '2pat_sort.bam'\n",
    "\n",
    "hap1mat = bs.AlignmentFile(p1mat, 'rb')\n",
    "hap2mat = bs.AlignmentFile(p2mat, 'rb')\n",
    "hap1pat = bs.AlignmentFile(p1pat, 'rb')\n",
    "hap2pat = bs.AlignmentFile(p2pat, 'rb')\n",
    "\n",
    "# Read Count\n",
    "\n",
    "p1mrc = int(pysam.view('-c', p1mat))\n",
    "p2mrc = int(pysam.view('-c', p2mat))\n",
    "p1prc = int(pysam.view('-c', p1pat))\n",
    "p2prc = int(pysam.view('-c', p2pat))\n",
    "\n",
    "# Maternal Comparisons \n",
    "m1 = []\n",
    "rm1 = []\n",
    "\n",
    "for i in range(p1mrc):\n",
    "    hap1matreads = next(hap1mat)\n",
    "    \n",
    "    a_seq = hap1matreads.query_sequence\n",
    "    \n",
    "    if a_seq is not None:\n",
    "        q_len = len(a_seq)\n",
    "    else:\n",
    "        q_len = 0\n",
    "        #print(bs.utils.flag_decode(hap1matreads.flag) )\n",
    "    m1.append([hap1matreads.read_name, hap1matreads.cigarstring, q_len])\n",
    "\n",
    "# Removes all reads that have an alignment length of \"0\"\n",
    "\n",
    "for i in range(len(m1)):\n",
    "    if m1[i][2] != 0:\n",
    "        rm1.append(m1[i])\n",
    "\n",
    "      \n",
    "m2= []\n",
    "rm2 = []\n",
    "\n",
    "for i in range(p2mrc):\n",
    "    \n",
    "    hap2matreads = next(hap2mat)\n",
    "    \n",
    "    a_seq = hap2matreads.query_sequence\n",
    "    \n",
    "    if a_seq is not None:\n",
    "        q_len = len(a_seq)\n",
    "    else:\n",
    "        q_len = 0    \n",
    "\n",
    "    m2.append([hap2matreads.read_name, hap2matreads.cigarstring, q_len])\n",
    "    \n",
    "for i in range(len(m2)):\n",
    "    if m2[i][2] != 0:\n",
    "        rm2.append(m2[i])\n",
    "        \n",
    "# Paternal Comparisons \n",
    "p1 = []\n",
    "rp1 = []\n",
    "\n",
    "for i in range(p1prc):\n",
    "    hap1patreads = next(hap1pat)\n",
    "    \n",
    "    a_seq = hap1patreads.query_sequence\n",
    "    \n",
    "    if a_seq is not None:\n",
    "        q_len = len(a_seq)\n",
    "    else:\n",
    "        q_len = 0\n",
    "        #print(bs.utils.flag_decode(hap1patreads.flag))\n",
    "        \n",
    "    p1.append([hap1patreads.read_name, hap1patreads.cigarstring, q_len])\n",
    "    \n",
    "for i in range(len(p1)):\n",
    "    if p1[i][2] != 0:\n",
    "        rp1.append(p1[i])\n",
    "\n",
    "p2 = []\n",
    "rp2 = []\n",
    "\n",
    "for i in range(p2prc):\n",
    "    hap2patreads = next(hap2pat)\n",
    "    \n",
    "    a_seq = hap2patreads.query_sequence\n",
    "    if a_seq is not None:\n",
    "        q_len = len(a_seq)\n",
    "    else:\n",
    "        q_len = 0\n",
    "        #print(bs.utils.flag_decode(hap2patreads.flag))    \n",
    "    p2.append([hap2patreads.read_name, hap2patreads.cigarstring, q_len])\n",
    "\n",
    "for i in range(len(p2)):\n",
    "    if p2[i][2] != 0:\n",
    "        rp2.append(p2[i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6d24eb45",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cigsort(cigstring):\n",
    "    matches = re.findall(r'(\\d+)([=SMIDHX]{1})', cigstring)\n",
    "    cigar_op = ([{'type':m[1], 'length':int(m[0])} for m in matches])\n",
    "    return cigar_op"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b4a8ce4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cigturn(cigdict, operation):\n",
    "    \n",
    "    #Segment of the query sequence that does not appear in the alignment. \n",
    "    seg = []\n",
    "    #Match (alignment column containing two letters). \n",
    "    match = []\n",
    "    #Insertion (gap in the query sequence). \n",
    "    ins = []\n",
    "    #Deletion (gap in the target sequence).\n",
    "    dele = []\n",
    "    #Segment of the query sequence that does not appear in the alignment. \n",
    "    nota = []\n",
    "    #Alignment column containing a mismatch, i.e. two different letters.\n",
    "    xma = []\n",
    "    # Alignment column containing two identical letters.\n",
    "    eq = []\n",
    "\n",
    "    for i in cigdict:\n",
    "        if i['type'] == 'S':\n",
    "            seg.append(i['length'])\n",
    "        elif i['type'] == 'M':\n",
    "            match.append(i['length'])\n",
    "        elif i['type'] == 'I':\n",
    "            ins.append(i['length'])\n",
    "        elif i['type'] == 'D':\n",
    "            dele.append(i['length'])\n",
    "        elif i['type'] == 'H':\n",
    "            nota.append(i['length'])\n",
    "        elif i['type'] == 'X':\n",
    "            xma.append(i['length'])\n",
    "        elif i['type'] == '=':\n",
    "            eq.append(i['length'])\n",
    "        \n",
    "\n",
    "    if operation == 'S':\n",
    "        return seg\n",
    "    elif operation == 'M':\n",
    "        return match\n",
    "    elif operation == 'I':\n",
    "        return ins\n",
    "    elif operation == 'D':\n",
    "        return dele\n",
    "    elif operation == 'H':\n",
    "        return nota\n",
    "    elif operation == 'X':\n",
    "        return xma\n",
    "    elif operation == '=':\n",
    "        return eq\n",
    "    else:\n",
    "        return [\"Inv\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a183f41f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Counting Matches Between Reads\n",
    "\n",
    "# Maternal Reads\n",
    "\n",
    "m1_match = []\n",
    "for i in range(len(rm1)):\n",
    "    m1mx = sum(cigturn(cigsort(rm1[i][1]),'X'))\n",
    "    m1sclip = sum(cigturn(cigsort(rm1[i][1]),'S'))\n",
    "    m1_match.append([rm1[i][0], m1mx/(rm1[i][2] - m1sclip)])\n",
    "    \n",
    "\n",
    "\n",
    "m2_match = []\n",
    "for i in range(len(rm2)):\n",
    "    m2mx = sum(cigturn(cigsort(rm2[i][1]),'X'))\n",
    "    m2sclip = sum(cigturn(cigsort(rm2[i][1]),'S'))\n",
    "    m2_match.append([rm2[i][0], m2mx/(rm2[i][2] - m2sclip)])\n",
    "    \n",
    "# Maternal Graph Parameters\n",
    "\n",
    "m_match = []\n",
    "mp1 = []\n",
    "mp2 = []\n",
    "mat_label = []\n",
    "\n",
    "for i in range(len(m1_match)):\n",
    "    for j in range(len(m2_match)):\n",
    "        if m1_match[i][0] == m2_match[j][0]:\n",
    "            m_match.append([m1_match[i][0], m1_match[i][1], m2_match[j][1]])\n",
    "            \n",
    "for i in range(len(m_match)):\n",
    "    mat_label.append(m_match[i][0])\n",
    "    mp1.append(m_match[i][1])\n",
    "    mp2.append(m_match[i][2])\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3bf4b047",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Counting Matches Between Reads\n",
    "\n",
    "# Paternal Reads\n",
    "\n",
    "p1_match = []\n",
    "for i in range(len(rp1)):\n",
    "    p1x = sum(cigturn(cigsort(rp1[i][1]),'X'))\n",
    "    p1sclip = sum(cigturn(cigsort(rp1[i][1]),'S'))\n",
    "    p1_match.append([rp1[i][0], p1x/(rp1[i][2] - p1sclip)])\n",
    "    \n",
    "p2_match = []\n",
    "\n",
    "for i in range(len(rp2)):\n",
    "    p2x = sum(cigturn(cigsort(rp2[i][1]),'X'))\n",
    "    p2sclip = sum(cigturn(cigsort(rp2[i][1]),'S'))\n",
    "    p2_match.append([rp2[i][0], p2x/(rp2[i][2] - p2sclip)])\n",
    "    \n",
    "# Paternal Graph Parameters\n",
    "\n",
    "p_match = []\n",
    "pp1 = []\n",
    "pp2 = []\n",
    "pat_label = []\n",
    "\n",
    "\n",
    "for i in range(len(p1_match)):\n",
    "    for j in range(len(p2_match)):\n",
    "        if p1_match[i][0] == p2_match[j][0]:\n",
    "            p_match.append([p1_match[i][0], p1_match[i][1], p2_match[j][1]])\n",
    "            #p_match.append([p1_match[i][0], p1_match[i][1], p2_match[j][1]])\n",
    "\n",
    "for i in range(len(p_match)):\n",
    "    pat_label.append(p_match[i][0])\n",
    "    pp1.append(p_match[i][1])\n",
    "    pp2.append(p_match[i][2])\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e52e02cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read Graphs Comparing Maternal and Paternal\n",
    "\n",
    "mp_label = []\n",
    "mp_len = []\n",
    "patp1 = []\n",
    "patp2 = []\n",
    "matp1 = []\n",
    "matp2 = []\n",
    "\n",
    "\n",
    "for j in range(len(mat_label)):\n",
    "    for i in range(len(pat_label)):\n",
    "        if pat_label[i] == mat_label[j]:\n",
    "            mp_label.append(mat_label[j])\n",
    "            patp1.append(pp1[i])\n",
    "            patp2.append(pp2[i])\n",
    "            matp1.append(mp1[j])\n",
    "            matp2.append(mp2[j])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3e6ce31d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Maternal Graph\n",
    "list_dict = {'Read Name': mat_label, 'Phased 1': mp1, 'Phased 2': mp2} \n",
    "df = pd.DataFrame(list_dict) \n",
    "df.to_csv('matp1p2.csv', index=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "93666d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Paternal Graph \n",
    "list_dict = {'Read Name': pat_label, 'Phased 1': pp1, 'Phased 2': pp2} \n",
    "df = pd.DataFrame(list_dict) \n",
    "df.to_csv('patp1p2.csv', index=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c667a43b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comparing Maternal Paternal Graph\n",
    "list_dict = {'Read Name': mp_label, 'Maternal Phased 1': matp1, 'Maternal Phased 2': matp2, 'Paternal Phased 1': patp1, 'Paternal Phased 2': patp2}\n",
    "df = pd.DataFrame(list_dict) \n",
    "df.to_csv('matpat.csv', index=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "277bceea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mismatch 404\n",
      "Soft clip 10579\n",
      "Hard clip 0\n",
      "Q align length 13224\n",
      "0\n",
      "10123S26=1X3=1X4=1X1=1X1=1X4=1X1=1X6=1X4=1X1=1X8=1X2=1X1=1X6=1X32=2X4=1X7=2X3=1X1=1X4=1X16=1X4=1I22=1X9=1X5=1X6=1X10=1X6=1X4=1X2=1X5=2X3=1X9=1X2=1I21=1X4=1X5=2X8=1X17=1X4=1X5=1I3=1X6=1X1=1X17=1X3=1X3=1X8=1X8=1X2=1X2=2X1=2X5=1X2=1X1=2X28=1X5=1X8=1X5=1X9=1I10=1X5=1D5=1X6=1X1=1X16=1X8=1X2=1X2=2I4=1X1=1X4=1X2=2X4=1X1=1X12=2X3=2X7=1X1=1X2=1X2=1X5=1X6=1X3=2X3=1X7=1X8=1X9=1X1=1X3=1X9=1X2=2X1=1D1X5=1D7=2X2=1X4=1X1=1X4=1X3=1X10=1X4=2X15=5D1X8=1X23=1X2=1X6=2X1=1D1X6=1X10=1D2X13=1X11=3X1=1X12=1X10=1X18=1X3=1X3=1X4=4D10=1X11=1X25=1X11=1X1=1X2=1X26=1X4=1X1=2X6=1X1=2X4=1X6=67I4=1X1=1X3=1X9=2X17=2X16=1X5=1X9=2X14=1X2=1X4=1X7=1X6=3X4=1X1=2X2=1X2=1X1=2X7=1X1=1X6=1I2=1X11=2X4=1X3=1X1=1X5=1X1=1X1=1X4=1X6=1X1=1X1=1X5=1X2=1X2=1I2=2X13=1X3=3X8=1D1X20=1X4=1I1=1X10=1X3=1X2=2X8=2X17=1X17=2X14=1X7=1X4=1X1=1X4=1X4=1X17=1X4=2X9=1X2=1X2=1X8=1X3=1X2=1X8=1X5=2X1=1X9=1X3=5I30=1X4=1X8=1X18=1X3=1X1=1X2=4X4=1X3=2X5=1X1=1X10=1X6=1X1=1X7=1X1=1X29=2X5=3X1=1X2=1X6=1X1=3X10=1X4=1X9=2X6=1X10=1X11=1X4=1X5=2X8=3X3=3I4=1X1=2X4=1X5=1X1=1X2=1X4=1X3=1X6=1X4=2X6=1X3=2X9=1X2=1X1=2X2=1X1=2X2=1X2=2X13=1X5=1X21=2X9=2X3=2X1=2X1=1X3=1I9=4I5=1X2=2X3=2X9=1X8=1X4=2X3=1D1=1X1=2X2=1X6=3I3=1X2=1X2=1X6=1X5=1X2=2D2=2D1X3=2X4=1X4=2X2=1X4=1X3=1X6=1X8=1X10=2X8=2X10=1X6=2X10=1X1=12I7=1X38=2X5=1X8=1X12=3X4=1X4=1D4=1X3=1X12=1I14=1X4=2X20=1X4=1X2=1X1=3D5=1X11=2X6=3X1=2X14=1I5=1X1=1X2=1X32=2X7=1X16=1X4=1X13=1X1=1X3=1X1=1I4=1X1=1X1=1X5=1X8=1X1=1X8=2X1=1X3=1I1=1X4=1X20=2X8=1X2=1X4=1X10=3X1=1X6=1X6=1X11=1X8=2X15=1X3=1X5=1X6=456S\n"
     ]
    }
   ],
   "source": [
    "p1pat = '1pat_sort.bam'\n",
    "hap1pat = bs.AlignmentFile(p1pat, 'rb')\n",
    "p1prc = int(pysam.view('-c', p1pat))\n",
    "\n",
    "for i in range(p1prc):\n",
    "    hap1patreads = next(hap1pat)\n",
    "    if hap1patreads.read_name == '0_phaseblock_93':\n",
    "        print(\"Mismatch\",sum(cigturn(cigsort(hap1patreads.cigarstring),'X')))\n",
    "        print(\"Soft clip\",sum(cigturn(cigsort(hap1patreads.cigarstring),'S')))\n",
    "        print(\"Hard clip\",sum(cigturn(cigsort(hap1patreads.cigarstring),'H')))\n",
    "        print(\"Q align length\",len(hap1patreads.query_sequence))\n",
    "        print(hap1patreads.flag)\n",
    "        print(hap1patreads.cigarstring)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a544269",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
